### ISSUE TITLE
On restart partition services take too long to start

### ISSUE THREAD

**Issue by deepthidevaki** [zeebe-io/zeebe#4158]
**Expected behavior**

Restart should be fast.

**Actual behavior**

On restart with existing data, partition service take long to start. It spents more time on loading segments.

<details><summary>Log</summary>
<p>

```
2020-01-13 15:32:56.140 [] [raft-server-2-system-partition-1] INFO  io.atomix.protocols.raft.partition.RaftPartitionGroup - Started
2020-01-13 15:32:56.213 [] [raft-partition-group-system-5] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=2, group=raft-partition}
2020-01-13 15:33:00.800 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 14 (raft-partition-partition-2-14.log)
2020-01-13 15:33:00.801 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 14 (raft-partition-partition-2-14.log)
2020-01-13 15:33:05.165 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 13 (raft-partition-partition-2-13.log)
2020-01-13 15:33:05.166 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 13 (raft-partition-partition-2-13.log)
2020-01-13 15:33:09.558 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 1 (raft-partition-partition-2-1.log)
2020-01-13 15:33:09.558 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 1 (raft-partition-partition-2-1.log)
2020-01-13 15:33:13.867 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 4 (raft-partition-partition-2-4.log)
2020-01-13 15:33:13.867 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 4 (raft-partition-partition-2-4.log)
2020-01-13 15:33:18.214 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 12 (raft-partition-partition-2-12.log)
2020-01-13 15:33:18.214 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 12 (raft-partition-partition-2-12.log)
2020-01-13 15:33:22.566 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 11 (raft-partition-partition-2-11.log)
2020-01-13 15:33:22.566 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 11 (raft-partition-partition-2-11.log)
2020-01-13 15:33:25.237 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 18 (raft-partition-partition-2-18.log)
2020-01-13 15:33:25.237 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 18 (raft-partition-partition-2-18.log)
2020-01-13 15:33:29.484 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 16 (raft-partition-partition-2-16.log)
2020-01-13 15:33:29.484 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 16 (raft-partition-partition-2-16.log)
2020-01-13 15:33:33.575 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 2 (raft-partition-partition-2-2.log)
2020-01-13 15:33:33.576 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 2 (raft-partition-partition-2-2.log)
2020-01-13 15:33:37.744 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 3 (raft-partition-partition-2-3.log)
2020-01-13 15:33:37.745 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 3 (raft-partition-partition-2-3.log)
2020-01-13 15:33:41.992 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 15 (raft-partition-partition-2-15.log)
2020-01-13 15:33:41.992 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 15 (raft-partition-partition-2-15.log)
2020-01-13 15:33:46.270 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 17 (raft-partition-partition-2-17.log)
2020-01-13 15:33:46.270 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 17 (raft-partition-partition-2-17.log)
2020-01-13 15:33:50.592 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 10 (raft-partition-partition-2-10.log)
2020-01-13 15:33:50.592 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 10 (raft-partition-partition-2-10.log)
2020-01-13 15:33:54.930 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 6 (raft-partition-partition-2-6.log)
2020-01-13 15:33:54.930 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 6 (raft-partition-partition-2-6.log)
2020-01-13 15:33:59.203 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 8 (raft-partition-partition-2-8.log)
2020-01-13 15:33:59.203 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 8 (raft-partition-partition-2-8.log)
2020-01-13 15:34:03.278 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 5 (raft-partition-partition-2-5.log)
2020-01-13 15:34:03.278 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 5 (raft-partition-partition-2-5.log)
2020-01-13 15:34:07.485 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 7 (raft-partition-partition-2-7.log)
2020-01-13 15:34:07.485 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 7 (raft-partition-partition-2-7.log)
2020-01-13 15:34:11.629 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 9 (raft-partition-partition-2-9.log)
2020-01-13 15:34:11.629 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 9 (raft-partition-partition-2-9.log)
2020-01-13 15:34:15.395 [] [raft-partition-group-system-5] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=1, group=raft-partition}
2020-01-13 15:34:15.395 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-2} - Transitioning to FOLLOWER
2020-01-13 15:34:15.403 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-2}{role=FOLLOWER} - Accepted VoteRequest{term=27, candidate=1, lastLogIndex=2076682, lastLogTerm=25}: candidate's log is up-to-date
2020-01-13 15:34:15.424 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-2}{role=FOLLOWER} - Accepted VoteRequest{term=33, candidate=1, lastLogIndex=2076682, lastLogTerm=25}: candidate's log is up-to-date
2020-01-13 15:34:15.509 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-2}{role=FOLLOWER} - Accepted VoteRequest{term=34, candidate=1, lastLogIndex=2076682, lastLogTerm=25}: candidate's log is up-to-date
2020-01-13 15:34:18.630 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-2}{role=FOLLOWER} - No heartbeat from null in the last PT3.117S (calculated from last 3117 ms)
2020-01-13 15:34:18.633 [] [raft-server-2-raft-partition-partition-2] WARN  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-2}{role=FOLLOWER} - io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
2020-01-13 15:34:18.951 [] [raft-server-2-raft-partition-partition-2] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-2} - Found leader 1
2020-01-13 15:34:19.612 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 17 (raft-partition-partition-1-17.log)
2020-01-13 15:34:19.613 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 17 (raft-partition-partition-1-17.log)
2020-01-13 15:34:23.799 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 3 (raft-partition-partition-1-3.log)
2020-01-13 15:34:23.799 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 3 (raft-partition-partition-1-3.log)
2020-01-13 15:34:28.565 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 10 (raft-partition-partition-1-10.log)
2020-01-13 15:34:28.566 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 10 (raft-partition-partition-1-10.log)
2020-01-13 15:34:33.009 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 8 (raft-partition-partition-1-8.log)
2020-01-13 15:34:33.009 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 8 (raft-partition-partition-1-8.log)
2020-01-13 15:34:37.480 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 13 (raft-partition-partition-1-13.log)
2020-01-13 15:34:37.480 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 13 (raft-partition-partition-1-13.log)
2020-01-13 15:34:42.000 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 11 (raft-partition-partition-1-11.log)
2020-01-13 15:34:42.001 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 11 (raft-partition-partition-1-11.log)
2020-01-13 15:34:46.365 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 5 (raft-partition-partition-1-5.log)
2020-01-13 15:34:46.366 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 5 (raft-partition-partition-1-5.log)
2020-01-13 15:34:50.775 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 6 (raft-partition-partition-1-6.log)
2020-01-13 15:34:50.775 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 6 (raft-partition-partition-1-6.log)
2020-01-13 15:34:54.986 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 2 (raft-partition-partition-1-2.log)
2020-01-13 15:34:54.987 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 2 (raft-partition-partition-1-2.log)
2020-01-13 15:34:59.041 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 4 (raft-partition-partition-1-4.log)
2020-01-13 15:34:59.041 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 4 (raft-partition-partition-1-4.log)
2020-01-13 15:35:03.505 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 14 (raft-partition-partition-1-14.log)
2020-01-13 15:35:03.510 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 14 (raft-partition-partition-1-14.log)
2020-01-13 15:35:06.720 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 18 (raft-partition-partition-1-18.log)
2020-01-13 15:35:06.721 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 18 (raft-partition-partition-1-18.log)
2020-01-13 15:35:10.988 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 1 (raft-partition-partition-1-1.log)
2020-01-13 15:35:10.989 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 1 (raft-partition-partition-1-1.log)
2020-01-13 15:35:15.424 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 7 (raft-partition-partition-1-7.log)
2020-01-13 15:35:15.425 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 7 (raft-partition-partition-1-7.log)
2020-01-13 15:35:19.978 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 9 (raft-partition-partition-1-9.log)
2020-01-13 15:35:19.978 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 9 (raft-partition-partition-1-9.log)
2020-01-13 15:35:24.413 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 15 (raft-partition-partition-1-15.log)
2020-01-13 15:35:24.414 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 15 (raft-partition-partition-1-15.log)
2020-01-13 15:35:28.730 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 12 (raft-partition-partition-1-12.log)
2020-01-13 15:35:28.731 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 12 (raft-partition-partition-1-12.log)
2020-01-13 15:35:33.079 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 16 (raft-partition-partition-1-16.log)
2020-01-13 15:35:33.080 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 16 (raft-partition-partition-1-16.log)
2020-01-13 15:35:37.578 [] [raft-partition-group-system-5] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=3, group=raft-partition}
2020-01-13 15:35:37.578 [] [raft-server-2-raft-partition-partition-1] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-1} - Transitioning to FOLLOWER
2020-01-13 15:35:39.806 [] [raft-server-2-raft-partition-partition-1] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Accepted PollRequest{term=8, candidate=1, lastLogIndex=2093495, lastLogTerm=8}: candidate's log is up-to-date
2020-01-13 15:35:39.829 [] [raft-server-2-raft-partition-partition-1] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Accepted VoteRequest{term=9, candidate=1, lastLogIndex=2093495, lastLogTerm=8}: candidate's log is up-to-date
2020-01-13 15:35:42.145 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 14 (raft-partition-partition-3-14.log)
2020-01-13 15:35:42.145 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 14 (raft-partition-partition-3-14.log)
2020-01-13 15:35:43.850 [] [raft-server-2-raft-partition-partition-1] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-1} - Found leader 1
2020-01-13 15:35:46.669 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 6 (raft-partition-partition-3-6.log)
2020-01-13 15:35:46.669 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 6 (raft-partition-partition-3-6.log)
2020-01-13 15:35:50.897 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 1 (raft-partition-partition-3-1.log)
2020-01-13 15:35:50.898 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 1 (raft-partition-partition-3-1.log)
2020-01-13 15:35:55.366 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 8 (raft-partition-partition-3-8.log)
2020-01-13 15:35:55.366 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 8 (raft-partition-partition-3-8.log)
2020-01-13 15:35:59.645 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 13 (raft-partition-partition-3-13.log)
2020-01-13 15:35:59.645 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 13 (raft-partition-partition-3-13.log)
2020-01-13 15:36:03.806 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 7 (raft-partition-partition-3-7.log)
2020-01-13 15:36:03.807 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 7 (raft-partition-partition-3-7.log)
2020-01-13 15:36:08.240 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 9 (raft-partition-partition-3-9.log)
2020-01-13 15:36:08.240 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 9 (raft-partition-partition-3-9.log)
2020-01-13 15:36:08.548 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 19 (raft-partition-partition-3-19.log)
2020-01-13 15:36:08.548 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 19 (raft-partition-partition-3-19.log)
2020-01-13 15:36:12.978 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 5 (raft-partition-partition-3-5.log)
2020-01-13 15:36:12.980 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 5 (raft-partition-partition-3-5.log)
2020-01-13 15:36:17.425 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 10 (raft-partition-partition-3-10.log)
2020-01-13 15:36:17.425 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 10 (raft-partition-partition-3-10.log)
2020-01-13 15:36:21.868 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 15 (raft-partition-partition-3-15.log)
2020-01-13 15:36:21.868 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 15 (raft-partition-partition-3-15.log)
2020-01-13 15:36:26.624 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 3 (raft-partition-partition-3-3.log)
2020-01-13 15:36:26.624 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 3 (raft-partition-partition-3-3.log)
2020-01-13 15:36:31.032 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 17 (raft-partition-partition-3-17.log)
2020-01-13 15:36:31.032 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 17 (raft-partition-partition-3-17.log)
2020-01-13 15:36:35.376 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 11 (raft-partition-partition-3-11.log)
2020-01-13 15:36:35.376 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 11 (raft-partition-partition-3-11.log)
2020-01-13 15:36:39.597 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 16 (raft-partition-partition-3-16.log)
2020-01-13 15:36:39.597 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 16 (raft-partition-partition-3-16.log)
2020-01-13 15:36:43.906 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 4 (raft-partition-partition-3-4.log)
2020-01-13 15:36:43.906 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 4 (raft-partition-partition-3-4.log)
2020-01-13 15:36:48.089 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 2 (raft-partition-partition-3-2.log)
2020-01-13 15:36:48.092 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 2 (raft-partition-partition-3-2.log)
2020-01-13 15:36:52.412 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 18 (raft-partition-partition-3-18.log)
2020-01-13 15:36:52.412 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 18 (raft-partition-partition-3-18.log)
2020-01-13 15:36:56.808 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Loaded disk segment: 12 (raft-partition-partition-3-12.log)
2020-01-13 15:36:56.809 [] [raft-partition-group-system-5] DEBUG io.atomix.storage.journal.SegmentedJournal - Found segment: 12 (raft-partition-partition-3-12.log)
2020-01-13 15:36:57.282 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-3} - Transitioning to CANDIDATE
2020-01-13 15:36:57.286 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.roles.CandidateRole - RaftServer{raft-partition-partition-3}{role=CANDIDATE} - Starting election
2020-01-13 15:36:57.309 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-3} - Transitioning to LEADER
2020-01-13 15:36:57.318 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-3} - Found leader 2
2020-01-13 15:36:57.782 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-3} - ConfigureRequest{term=4, leader=2, index=0, timestamp=1578925341633, members=[DefaultRaftMember{id=0, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=2, type=BOOTSTRAP, updated=2020-01-13T14:22:21.633854Z}]} to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
2020-01-13 15:36:57.786 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-3} - ConfigureRequest{term=4, leader=2, index=0, timestamp=1578925341633, members=[DefaultRaftMember{id=0, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=2, type=BOOTSTRAP, updated=2020-01-13T14:22:21.633854Z}]} to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
2020-01-13 15:36:57.800 [] [raft-server-2-raft-partition-partition-3] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-3} - ConfigureRequest{term=4, leader=2, index=0, timestamp=1578925341633, members=[DefaultRaftMember{id=0, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-01-13T14:22:21.633854Z}, DefaultRaftMember{id=2, type=BOOTSTRAP, updated=2020-01-13T14:22:21.633854Z}]} to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
2020-01-13 15:37:09.100 [] [raft-partition-group-raft-partition-0] INFO  io.atomix.protocols.raft.partition.RaftPartitionGroup - Started
2020-01-13 15:37:09.101 [] [raft-partition-group-raft-partition-0] INFO  io.atomix.primitive.partition.impl.DefaultPartitionService - Started
2020-01-13 15:37:09.239 [] [raft-partition-group-system-3] INFO  io.atomix.core.impl.CoreTransactionService - Started
2020-01-13 15:37:09.240 [] [raft-partition-group-system-3] INFO  io.atomix.core.impl.CorePrimitivesService - Started
2020-01-13 15:37:09.241 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-2 [6/10]: cluster services started in 263205 ms
```
</p></details>

**Steps to reproduce**

Restart a cluster with many log segments in one partition.

**Environment**

 - Atomix: [3.2.0-alpha10]

**Comment by kjozsa** [comment]
We even experienced restart times longer than 15 minutes, probably hitting the same issue. Is this problem scheduled to get resolved for 0.23.0?

**Comment by andylholloway** [comment]
We are also seeing this; the bootstrap time can take over 20 mins to complete. 
```
2020-04-14 10:49:53.525 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [3/10]: command api transport started in 416 ms
2020-04-14 10:49:53.525 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [4/10]: command api handler
2020-04-14 10:49:53.617 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [4/10]: command api handler started in 91 ms
2020-04-14 10:49:53.618 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [5/10]: subscription api
2020-04-14 10:49:53.810 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [5/10]: subscription api started in 192 ms
2020-04-14 10:49:53.811 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [6/10]: cluster services
2020-04-14 11:11:14.518 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [6/10]: cluster services started in 1280707 ms
2020-04-14 11:11:14.519 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [7/10]: topology manager
2020-04-14 11:11:14.520 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [7/10]: topology manager started in 1 ms
2020-04-14 11:11:14.521 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [8/10]: metric's server
2020-04-14 11:11:14.531 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [8/10]: metric's server started in 10 ms
2020-04-14 11:11:14.532 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [9/10]: leader management request handler
2020-04-14 11:11:14.533 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [9/10]: leader management request handler started in 1 ms
2020-04-14 11:11:14.534 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [10/10]: zeebe partitions
2020-04-14 11:11:14.536 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 partitions [1/1]: partition 1
2020-04-14 11:11:14.881 [] [main] DEBUG io.zeebe.broker.exporter - Exporter configured with ElasticsearchExporterConfiguration{url='http://elasticsearch-master:9200', index=IndexConfiguration{indexPrefix='zeebe-rec
ord', createTemplate=true, command=false, event=true, rejection=false, error=true, deployment=true, incident=true, job=true, message=false, messageSubscription=false, variable=true, variableDocument=true, workflowI
nstance=true, workflowInstanceCreation=false, workflowInstanceSubscription=false}, bulk=BulkConfiguration{delay=5, size=1000}, authentication=AuthenticationConfiguration{username='null'}}
2020-04-14 11:11:15.051 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Removing follower partition service for partition PartitionId{id=1, group=raft-partition}
2020-04-14 11:11:15.115 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Partition role transitioning from null to LEADER
2020-04-14 11:11:15.115 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Installing leader partition service for partition PartitionId{id=1, group=raft-partition}
2020-04-14 11:11:15.532 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=38655128624, path=/usr/local/zeebe/data/raft-partition/par
titions/1/snapshots/6054-38-1586267129817-38655128624}, SnapshotImpl{position=38655109208, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/5997-38-1586266229810-38655109208}, SnapshotImpl{position=
38655094760, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/5955-38-1586265329782-38655094760}]
2020-04-14 11:11:16.468 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/1/runtime'.
2020-04-14 11:11:16.470 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=38655128624, path=/usr/local/zeebe/data/raft-part
ition/partitions/1/snapshots/6054-38-1586267129817-38655128624}'
2020-04-14 11:11:16.722 [Broker-0-LogStream-1] [Broker-0-zb-actors-0] DEBUG io.zeebe.logstreams - Configured log appender back pressure at partition 1 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, al
phaLimit=0.7, betaLimit=0.95}. Window limiting is disabled
2020-04-14 11:11:16.871 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-0] DEBUG io.zeebe.logstreams - Recovering state of partition 1 from snapshot
2020-04-14 11:11:17.058 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-0] INFO  io.zeebe.logstreams - Recovered state of partition 1 from snapshot at position 38655128624
2020-04-14 11:11:17.852 [Broker-0-SnapshotDirector-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - The position of the last valid snapshot is '38655128624'. Taking snapshots beyond this position.
2020-04-14 11:11:17.915 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Recovering exporter from snapshot
2020-04-14 11:11:17.920 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 partitions [1/1]: partition 1 started in 3383 ms
2020-04-14 11:11:17.920 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 partitions succeeded. Started 1 steps in 3384 ms.
2020-04-14 11:11:17.920 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [10/10]: zeebe partitions started in 3386 ms
2020-04-14 11:11:17.920 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 succeeded. Started 10 steps in 1292283 ms.
2020-04-14 11:11:17.924 [Broker-0-HealthCheckService] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - All partitions are installed. Broker is ready!
2020-04-14 11:11:18.211 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Recovered exporter 'Broker-0-Exporter-1' from snapshot at lastExportedPosition 38655128624
2020-04-14 11:11:18.212 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Configure exporter with id 'elasticsearch'
2020-04-14 11:11:18.212 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter.elasticsearch - Exporter configured with ElasticsearchExporterConfiguration{url='http://elasticsearch-master:9
200', index=IndexConfiguration{indexPrefix='zeebe-record', createTemplate=true, command=false, event=true, rejection=false, error=true, deployment=true, incident=true, job=true, message=false, messageSubscription=f
alse, variable=true, variableDocument=true, workflowInstance=true, workflowInstanceCreation=false, workflowInstanceSubscription=false}, bulk=BulkConfiguration{delay=5, size=1000}, authentication=AuthenticationConfi
guration{username='null'}}
```

We have seen this in 2 separate development clusters, both clusters are fairly new so contain only a small amount of data (i.e. a low number of workflow instances have been created, less then 100).

Cluster:
cluster size: 1
partition count: 1
replication: 1

Exporters:
ElasticSearch
Hazelcast

**Comment by kjozsa** [comment]
@andylholloway using which zeebe version?

**Comment by andylholloway** [comment]
@kjozsa we are using version 0.22.1.

**Comment by kjozsa** [comment]
So this no longer happens with 0.23.1? We're yet to upgrade to this version..

**Comment by npepinpe** [comment]
No, this is still an issue, there were some problem with the bot configuration :sweat_smile:

**Comment by npepinpe** [comment]
Superseded by #3028

**Comment by npepinpe** [comment]
Reopening as #3028 can be used as more of a milestone issue from which others are created.

**Comment by deepthidevaki** [comment]
I have been looking into this issue. 
With 512MB segment size it takes ~4s to load a segment. So the load time will increase linearly with the number of segments (total over all partitions). See the duration between loading two segments: 

```
D 2020-07-06T09:50:25.591089Z Loaded disk segment: 2 (raft-partition-partition-1-2.log) 
D 2020-07-06T09:50:29.107726Z Loaded disk segment: 1 (raft-partition-partition-1-1.log) 
D 2020-07-06T09:50:32.661657Z Loaded disk segment: 3 (raft-partition-partition-1-3.log) 
D 2020-07-06T09:50:35.878432Z Loaded disk segment: 4 (raft-partition-partition-1-4.log) 
```

In a previous issue, we had refactored the read operation in `FileChannelReader` to make the read more efficient. I applied similar logic here to read the entries. It doesn't yield a visible improvement. (Probably need a microbenchmark to see if it has any real impact.)
```
D 2020-07-07T09:46:38.412023Z Loaded disk segment: 7 (raft-partition-partition-1-7.log) 
D 2020-07-07T09:46:42.469432Z Loaded disk segment: 5 (raft-partition-partition-1-5.log) 
D 2020-07-07T09:46:46.576228Z Loaded disk segment: 4 (raft-partition-partition-1-4.log) 
D 2020-07-07T09:46:48.502805Z Loaded disk segment: 8 (raft-partition-partition-1-8.log) 
D 2020-07-07T09:46:52.522026Z Loaded disk segment: 6 (raft-partition-partition-1-6.log
```

Reduced maxMessageSize to 1MB to see if cache miss is the main reason. It improved the load time slightly to ~3s.
```
D 2020-07-07T09:26:49.496186Z Loaded disk segment: 1 (raft-partition-partition-1-1.log) 
D 2020-07-07T09:26:53.079448Z Loaded disk segment: 4 (raft-partition-partition-1-4.log) 
D 2020-07-07T09:26:56.996421Z Loaded disk segment: 6 (raft-partition-partition-1-6.log) 
D 2020-07-07T09:27:00.408339Z Loaded disk segment: 5 (raft-partition-partition-1-5.log) 
D 2020-07-07T09:27:00.469793Z Loaded disk segment: 8 (raft-partition-partition-1-8.log) 
D 2020-07-07T09:27:03.783912Z Loaded disk segment: 2 (raft-partition-partition-1-2.log) 
D 2020-07-07T09:27:07.082357Z Loaded disk segment: 3 (raft-partition-partition-1-3.log) 
D 2020-07-07T09:27:10.336075Z Loaded disk segment: 7 (raft-partition-partition-1-7.log) 
```

**Comment by deepthidevaki** [comment]
At startup, raft partition loads all segments and does the following for each segment :
	* Read each entry
	* Calculate and verify checksum
	* Update JournalIndex (Both zeebeposition-to-index and index-to-bufferposition)
	* Record the lastEntry of the segment.
	
To reduce the load time, we have to avoid reading the whole segment. 
* Not verifying checksum is ok, as we do it during the read (to be verified).
* Not updating index might increase the lookup time. If sparse index is acceptable, then we can avoid this step.
* `lastEntry` should be updated because otherwise it might break the system. 

So we need some way to updated `lastEntry` at startup without scanning the whole segment. For that I added  "lastEntryIndex" and "lastEntryPosition" in journal descriptor and persist it in the segment. So on startup, it can directly go to the "lastEntryPosition" and read the lastEntry. With these changes the cluster is running fine and the load time is negligible:

```
D 2020-07-07T14:40:52.448265Z Loaded disk segment: 17 (raft-partition-partition-1-17.log) 
D 2020-07-07T14:40:52.455337Z Loaded disk segment: 19 (raft-partition-partition-1-19.log) 
D 2020-07-07T14:40:52.473090Z Loaded disk segment: 23 (raft-partition-partition-1-23.log) 
D 2020-07-07T14:40:52.484288Z Loaded disk segment: 22 (raft-partition-partition-1-22.log) 
D 2020-07-07T14:40:52.491106Z Loaded disk segment: 20 (raft-partition-partition-1-20.log) 
D 2020-07-07T14:40:52.517089Z Loaded disk segment: 21 (raft-partition-partition-1-21.log) 
D 2020-07-07T14:40:52.523861Z Loaded disk segment: 18 (raft-partition-partition-1-18.log) 
D 2020-07-07T14:40:56.730822Z Loaded disk segment: 24 (raft-partition-partition-1-24.log) 
```

The issue with this approach is that the index is not updated on restart. Index will be sparse and contains only first and last entry in a segment (except for the last segment). This will impact the lookup time. I don't know yet how bad this is. Another solution is  to persist the index so that we don't have to rebuild it on restart.

**Comment by deepthidevaki** [comment]
Also found out that on startup, raft server reads and applies all entries in the log. Only once it has applied until the committed index the server is ready. On the leader it is initiated after the initial entry is committed.

```
I 2020-07-08T13:02:01.109066Z RaftServer{raft-partition-partition-3}{role=LEADER} - Initial entry committed 
I 2020-07-08T13:03:26.099914Z RaftServer{raft-partition-partition-3}{role=LEADER} - Initial entry applied 
```
On the follower it is initiated when when it gets the first appendRequest with a positive commit index.

In ZeebeRaftStateMachine, apply(index) does a applyAll which reads every entry in the logs and applies it. So I removed this `applyAll` and set it to only apply the given index.

Before this change:
```
I 2020-07-08T13:51:39.178518Z Bootstrapping server for partition PartitionId{id=3, group=raft-partition} 
I 2020-07-08T13:51:39.180056Z RaftServer{raft-partition-partition-3} - Transitioning to FOLLOWER 
I 2020-07-08T13:51:39.195442Z RaftServer{raft-partition-partition-3} - Successfully joined via 1 
I 2020-07-08T13:51:39.206321Z RaftServer{raft-partition-partition-3} - Found leader 1 
I 2020-07-08T13:53:07.699345Z RaftServer{raft-partition-partition-3} - Server started successfully! 
I 2020-07-08T13:53:07.699991Z Successfully started server for partition PartitionId{id=3, group=raft-partition} 
....
I 2020-07-08T13:53:08.028694Z Bootstrap Broker-0 succeeded. Started 10 steps in 114737 ms. 
```

After:
```
I 2020-07-08T13:57:35.326926Z Bootstrapping server for partition PartitionId{id=3, group=raft-partition} 
I 2020-07-08T13:57:35.328108Z RaftServer{raft-partition-partition-3} - Transitioning to FOLLOWER 
I 2020-07-08T13:57:35.334699Z RaftServer{raft-partition-partition-3} - Successfully joined via 2 
I 2020-07-08T13:57:35.459526Z RaftServer{raft-partition-partition-3} - Found leader 1 
I 2020-07-08T13:57:38.056188Z RaftServer{raft-partition-partition-3} - Server started successfully! 
I 2020-07-08T13:57:38.057105Z Successfully started server for partition PartitionId{id=3, group=raft-partition} 
....
I 2020-07-08T13:57:38.398001Z Bootstrap Broker-0 succeeded. Started 10 steps in 26143 ms. 
```

This applying entries was required originally to build the in-memory state of the primitives. Since we are building the state differently (via StreamProcessors), I would challenge if we really need to keep this.

Btw this is also a reason for OOM observed during the restart in some cases. @pihme is already trying to confirm it and prevent OOM due to this without removing this behavior.

**Comment by Zelldon** [comment]
I think we can remove this yes.

**Comment by npepinpe** [comment]
We're reprocessing anyway so I don't see any point in applying all entries, especially when our ZeebePartition is not installed. We should check that removing this applyAll is not useful during normal run, but otherwise we could remove it.

Great catch :slightly_smiling_face:

**Comment by deepthidevaki** [comment]
We use this to notifyCommitListeners. When an index is applied, the entry at that index is send to the commit listeners. LogStream uses this to record the commit position from a ZeebeEntry. If we remove apply we need some way to reliable notify commits. May be notifying from the LeaderRole after an entry is committed is enough. Not sure if we break any edge cases - such as during failover.

**Comment by npepinpe** [comment]
As discussed, please summarize final status and close :+1:

**Comment by deepthidevaki** [comment]
In the latest tests, for broker with 100GB full of log segments around 70*3 log segments in total, it took around 13 minutes to restart. We don't expect a broker to restart with so many segments often. A broker can end up in such a state only in very specific cases like the exporter service is not available for a long time (for hours or days). 
To reduce the startup time, one idea is not to scan the log at the beginning. But scanning is necessary to verify the checksums of the segments. Hence we don't see an immediately solution for this. 

Also the long restart time reported by users were due to other issues (most of them fixed by now), and not due to large number of log segments. So I would assume this is not a normal scenario and no users are affected by this so far.

We should in general put effort in improving our journal performance. Since we don't plan to work on this issue any more let's close this issue. If we find this is affecting any users, we can revisit this.
